### PR TITLE
[WIP] Thumb2

### PR DESCRIPTION
**Pull request by sunho**
Hello! I'm from Vita3k. I want to use dynarmic in our emulator, but ps vita applications heavily use thumb1 and thumb2 code. Dynarmic lacks thumb2 support, so I'm adding it in this pr!

### CODE REVIEW

File: src/frontend/A32/ITState.h [hunk 1]
```diff
@@ -22,31 +22,39 @@ class ITState final {
...
     }
 
     bool IsInITBlock() const {
-        return Mask() != 0b0000;
+        return Flags() != 0b00000;
     }
-
     bool IsLastInITBlock() const {
-        return Mask() == 0b1000;
+        return Common::Bits<1,4>(Flags()) == 0b1000;
```
Comment (lioncash) [inline src/frontend/A32/ITState.h:51]:
```suggestion
        return Common::Bits<1, 4>(Flags()) == 0b1000;
```

File: src/frontend/A32/disassembler/disassembler_thumb.cpp [hunk 2]
```diff
@@ -269,6 +269,30 @@ class DisassemblerVisitor {
...
+    std::string thumb16_IT(Imm<4> firstcond, Imm<4> mask) {
+        int until = -1;
+        std::string out = "it";
+        for (int i = 0; i < 4; i++) {
+            if (Common::Bit(i, mask.ZeroExtend())) {
+                until = i;
+                break;
+            }
+        }
+        if (until == -1)
```
Comment (lioncash) [inline src/frontend/A32/disassembler/disassembler_thumb.cpp:281]: This should be braced.

File: src/frontend/A32/disassembler/disassembler_thumb.cpp [hunk 3]
```diff
@@ -269,6 +269,30 @@ class DisassemblerVisitor {
...
+            return "undefined";
+        for (int i = 3; i > until; i--) {
+            if (Common::Bit(i, mask.ZeroExtend()) == firstcond.Bit<0>()) {
+                out += "t";
+            } else {
+                out += "e";
+            }
+        }
+        out += " ";
```
Comment (lioncash) [inline src/frontend/A32/disassembler/disassembler_thumb.cpp:290]:
```suggestion
        out += ' ';
```

File: src/frontend/A32/disassembler/disassembler_thumb.cpp [hunk 4]
```diff
@@ -269,6 +269,30 @@ class DisassemblerVisitor {
...
+            if (Common::Bit(i, mask.ZeroExtend())) {
+                until = i;
+                break;
+            }
+        }
+        if (until == -1)
+            return "undefined";
+        for (int i = 3; i > until; i--) {
+            if (Common::Bit(i, mask.ZeroExtend()) == firstcond.Bit<0>()) {
+                out += "t";
```
Comment (lioncash) [inline src/frontend/A32/disassembler/disassembler_thumb.cpp:285]:
```suggestion
                out += 't';
```

File: src/frontend/A32/disassembler/disassembler_thumb.cpp [hunk 5]
```diff
@@ -269,6 +269,30 @@ class DisassemblerVisitor {
...
+            } else {
+                out += "e";
```
Comment (lioncash) [inline src/frontend/A32/disassembler/disassembler_thumb.cpp:287]:
```suggestion
                out += 'e';
```

File: src/frontend/A32/translate/impl/thumb16.cpp [hunk 6]
```diff
@@ -943,11 +1227,43 @@ bool ThumbTranslatorVisitor::thumb16_B_t1(Cond cond, Imm<8> imm8) {
...
+    }
+    if (firstcond == Cond::NV) {
+       // NV conditional is obsolete
+       return UnpredictableInstruction();
+    }
+    if (mask == 0b0000) {
+        return UndefinedInstruction();
+    }
+    
+    auto new_it = ITState(0);
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/thumb16.cpp:1254]: These variables should be marked as const where applicable.

File: src/frontend/A32/translate/impl/thumb32.cpp [hunk 7]
```diff
@@ -35,6 +43,321 @@ bool ThumbTranslatorVisitor::thumb32_BLX_imm(Imm<11> hi, Imm<11> lo) {
...
+    }
+    if (n == Reg::PC || d == Reg::PC) {
+        return UnpredictableInstruction();
+    }
+    
+    const auto cpsr_c = ir.GetCFlag();
+    const auto imm_carry = ThumbExpandImm_C(i, imm3, imm8, cpsr_c);
+    const auto result = ir.And(ir.GetRegister(n), ir.Imm32(imm_carry.imm32));
+
+
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/thumb32.cpp:110]: Unnecessary newline

File: src/frontend/A32/translate/impl/thumb32.cpp [hunk 8]
```diff
@@ -35,6 +43,321 @@ bool ThumbTranslatorVisitor::thumb32_BLX_imm(Imm<11> hi, Imm<11> lo) {
...
+    }
+    if (n == Reg::PC || d == Reg::PC) {
+        return UnpredictableInstruction();
+    }
+    
+    const auto cpsr_c = ir.GetCFlag();
+    const auto imm_carry = ThumbExpandImm_C(i, imm3, imm8, cpsr_c);
+    const auto result = ir.Or(ir.GetRegister(n), ir.Imm32(imm_carry.imm32));
+
+
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/thumb32.cpp:133]: Unnecessary newline

File: src/frontend/A32/translate/impl/translate_thumb.h [hunk 9]
```diff
@@ -22,10 +33,50 @@ struct ThumbTranslatorVisitor final {
     explicit ThumbTranslatorVisitor(IR::Block& block, LocationDescriptor descriptor, const TranslationOptions& options) : ir(block, descriptor), options(options) {
         ASSERT_MSG(descriptor.TFlag(), "The processor must be in Thumb mode");
     }
+    
+    static u32 ThumbExpandImm(Imm<1> i, Imm<3> imm3, Imm<8> imm8) {
+        const auto imm12 = concatenate(i, imm3, imm8);
+        if (imm12.Bits<10, 11>() == 0) {
+            u32 imm8 = imm12.Bits<0, 7>();
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/translate_thumb.h:40]:
```suggestion
            const u32 imm8 = imm12.Bits<0, 7>();
```

File: src/frontend/A32/translate/impl/translate_thumb.h [hunk 10]
```diff
@@ -22,10 +33,50 @@ struct ThumbTranslatorVisitor final {
...
+                case 0b01:
+                    return (imm8 << 16) | imm8;
+                case 0b10:
+                    return (imm8 << 24) | (imm8 << 8);
+                case 0b11:
+                    return Common::Replicate(imm8, 8);
+            }
+            assert(false);
+        }
+        int rotate = imm12.Bits<7, 11>();
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/translate_thumb.h:53]:
```suggestion
        const int rotate = imm12.Bits<7, 11>();
```

File: src/frontend/A32/translate/impl/translate_thumb.h [hunk 11]
```diff
@@ -22,10 +33,50 @@ struct ThumbTranslatorVisitor final {
...
+        Imm<8> unrotated_value = concatenate(Imm<1>(1), Imm<7>(imm12.Bits<0, 6>()));
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/translate_thumb.h:54]:
```suggestion
        const Imm<8> unrotated_value = concatenate(Imm<1>(1), Imm<7>(imm12.Bits<0, 6>()));
```

File: src/frontend/A32/translate/impl/translate_thumb.h [hunk 12]
```diff
@@ -22,10 +33,50 @@ struct ThumbTranslatorVisitor final {
...
+        return Common::RotateRight<u32>(unrotated_value.ZeroExtend(), rotate);
+    }
+    
+    struct ImmAndCarry {
+        u32 imm32;
+        IR::U1 carry;
+    };
+    
+    ImmAndCarry ThumbExpandImm_C(Imm<1> i, Imm<3> imm3, Imm<8> imm8, IR::U1 carry_in) {
+        u32 imm32 = ThumbExpandImm(i, imm3, imm8);
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/translate_thumb.h:64]:
```suggestion
        const u32 imm32 = ThumbExpandImm(i, imm3, imm8);
```

File: src/frontend/A32/translate/translate_thumb.cpp [hunk 13]
```diff
@@ -54,6 +54,15 @@ std::tuple<u32, ThumbInstSize> ReadThumbInstruction(u32 arm_pc, MemoryReadCodeFu
 
 } // local namespace
 
+static bool CondCanContinue(ConditionalState cond_state, const A32::IREmitter& ir) {
+    ASSERT_MSG(cond_state != ConditionalState::Break, "Should never happen.");
+    if (cond_state == ConditionalState::None)
```
Comment (lioncash) [inline src/frontend/A32/translate/translate_thumb.cpp:59]: This should be braced

File: tests/A32/fuzz_thumb.cpp [hunk 14]
```diff
@@ -242,33 +272,126 @@ TEST_CASE("Fuzz Thumb instructions set 1", "[JitX64][Thumb]") {
...
+        Thumb32InstGen("11110i01101Snnnn0kkkddddmmmmmmmm", // ADD (imm)
+                     [](u32 inst){ return Common::Bits<8, 11>(inst) != 0b1111 && Common::Bits<16, 19>(inst) != 0b1111; }),
+            // R15 is UNPREDICTABLE
+        Thumb32InstGen("11110i01110Snnnn0kkkddddmmmmmmmm", // RSB (imm)
+                 [](u32 inst){ return Common::Bits<8, 11>(inst) != 0b1111 && Common::Bits<16, 19>(inst) != 0b1111; }),
+            // R15 is UNPREDICTABLE
+    };
+
+    const auto instruction_select = [&](int) -> u32 {
+        size_t inst_index = RandInt<size_t>(0, instructions.size() - 1);
```
Comment (lioncash) [inline tests/A32/fuzz_thumb.cpp:380]:
```suggestion
        const size_t inst_index = RandInt<size_t>(0, instructions.size() - 1);
```

File: tests/A32/fuzz_thumb.cpp [hunk 15]
```diff
@@ -242,33 +272,126 @@ TEST_CASE("Fuzz Thumb instructions set 1", "[JitX64][Thumb]") {
...
+        Thumb16InstGen("010001100hxxxxxx") // MOV (high registers)6 inst)
+    };
+
+    const auto cond_instruction_select = [&](int i) -> u16 {
+        if (i == 2 || i == 7) {
+            // IT
+            return Thumb16InstGen("10111111cccc1mmm",
+              [](u16 inst){ return Common::Bits<4, 7>(inst) != 0b1111 && Common::Bits<4, 7>(inst) != 0b1110; }).Generate();
+        }
+        size_t inst_index = RandInt<size_t>(0, instructions.size() - 1);
```
Comment (lioncash) [inline tests/A32/fuzz_thumb.cpp:325]:
```suggestion
        const size_t inst_index = RandInt<size_t>(0, instructions.size() - 1);
```

File: tests/A32/fuzz_thumb.cpp [hunk 16]
```diff
@@ -195,42 +200,67 @@ void FuzzJitThumb(const size_t instruction_count, const size_t instructions_to_e
...
+    A32Unicorn uni{test_env};
+    A32::Jit jit{GetUserConfig(&test_env)};
 
+    for (size_t run_number = 0; run_number < run_count; run_number++) {
+        ThumbTestEnv::RegisterArray initial_regs;
+        std::generate_n(initial_regs.begin(), initial_regs.size() - 1, []{ return RandInt<u32>(0, 0xFFFFFFFF); });
+        initial_regs[15] = 0;
+        
+        for (size_t i = 0; i < instruction_count; i++) {
+            auto insn = instruction_generator(i);
```
Comment (lioncash) [inline tests/A32/fuzz_thumb.cpp:226]:
```suggestion
            const auto insn = instruction_generator(i);
```

File: src/frontend/A32/disassembler/disassembler_thumb.cpp [hunk 17]
```diff
@@ -278,16 +278,17 @@ class DisassemblerVisitor {
                 break;
             }
         }
-        if (until == -1)
-            return "undefined";
-        for (int i = 3; i > until; i--) {
+		if (until == -1) {
```
Comment (lioncash) [inline src/frontend/A32/disassembler/disassembler_thumb.cpp:281]: We use spaces, not tabs for indentation

File: src/frontend/A32/translate/translate_thumb.cpp [hunk 18]
```diff
@@ -56,8 +56,9 @@ std::tuple<u32, ThumbInstSize> ReadThumbInstruction(u32 arm_pc, MemoryReadCodeFu
 
 static bool CondCanContinue(ConditionalState cond_state, const A32::IREmitter& ir) {
     ASSERT_MSG(cond_state != ConditionalState::Break, "Should never happen.");
-    if (cond_state == ConditionalState::None)
-        return true;
+	if (cond_state == ConditionalState::None) {
+		return true;
```
Comment (lioncash) [inline src/frontend/A32/translate/translate_thumb.cpp:60]: This is using tabs for indentation when it should be spaces.

File: src/frontend/A32/disassembler/disassembler_thumb.cpp (hunk 17 above)
Comment (sunho) [reply src/frontend/A32/disassembler/disassembler_thumb.cpp:281]: Opps

File: src/frontend/A32/translate/impl/helper.cpp [hunk 19]
```diff
@@ -0,0 +1,49 @@
+/* This file is part of the dynarmic project.
+ * Copyright (c) 2016 MerryMage
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/helper.cpp:2]:
```suggestion
 * Copyright (c) 2020 MerryMage
```

File: src/frontend/A32/translate/helper.h [hunk 20]
```diff
@@ -0,0 +1,17 @@
+/* This file is part of the dynarmic project.
+ * Copyright (c) 2016 MerryMage
```
Comment (lioncash) [inline src/frontend/A32/translate/helper.h:2]:
```suggestion
 * Copyright (c) 2020 MerryMage
```

File: src/frontend/A32/translate/helper.h [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+#include "common/assert.h"
+#include "frontend/A32/ir_emitter.h"
+#include "frontend/A32/types.h"
+
+namespace Dynarmic::A32::Helper {
+
+bool LDMHelper(A32::IREmitter& ir, bool W, Reg n, RegList list, IR::U32 start_address, IR::U32 writeback_address);
+bool STMHelper(A32::IREmitter& ir, bool W, Reg n, RegList list, IR::U32 start_address, IR::U32 writeback_address);
+
+} // namespace Dynarmic::A32
```
Comment (lioncash) [inline src/frontend/A32/translate/helper.h:17]:
```suggestion
} // namespace Dynarmic::A32::Helper
```

File: src/frontend/A32/translate/helper.h [hunk 22]
```diff
@@ -0,0 +1,17 @@
+/* This file is part of the dynarmic project.
+ * Copyright (c) 2016 MerryMage
+ * SPDX-License-Identifier: 0BSD
+ */
+
+#pragma once
+
+#include "common/assert.h"
```
Comment (lioncash) [inline src/frontend/A32/translate/helper.h:8]: This include isn't necessary here.

File: src/frontend/A32/translate/impl/helper.cpp [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+ * SPDX-License-Identifier: 0BSD
+ */
+
+#include "frontend/A32/translate/helper.h"
+
+namespace Dynarmic::A32::Helper {
+
+bool LDMHelper(A32::IREmitter& ir, bool W, Reg n, RegList list, IR::U32 start_address, IR::U32 writeback_address) {
+	auto address = start_address;
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/helper.cpp]: We don't use tabs for indentation, we use spaces. Ditto for the rest of the file.

File: src/frontend/A32/translate/impl/helper.cpp [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+			ir.SetRegister(static_cast<Reg>(i), ir.ReadMemory32(address));
+			address = ir.Add(address, ir.Imm32(4));
+		}
+	}
+	if (W && !Common::Bit(RegNumber(n), list)) {
+		ir.SetRegister(n, writeback_address);
+	}
+	if (Common::Bit<15>(list)) {
+		ir.LoadWritePC(ir.ReadMemory32(address));
+		if (n == Reg::R13)
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/helper.cpp]: This and the else condition should be braced.

File: src/frontend/A32/translate/impl/thumb32.cpp [hunk 25]
```diff
@@ -356,6 +356,95 @@ bool ThumbTranslatorVisitor::thumb32_RSB_imm(Imm<1> i, bool S, Reg n, Imm<3> imm
     return true;
 }
 
+// STMIA<c>.W <Rn>{!},<registers>
+bool ThumbTranslatorVisitor::thumb32_STMIA(bool W, Reg n, RegList reg_list) {
+	if (!ConditionPassed()) {
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/thumb32.cpp]: We don't use tabs for indentation. This applies to all additions to this file.

File: src/frontend/A32/translate/impl/translate_thumb.h [hunk 26]
```diff
@@ -164,6 +165,11 @@ struct ThumbTranslatorVisitor final {
     bool thumb16_IT(Cond firstcond, Imm<4> mask);
 
     // thumb32
+	bool thumb32_STMIA(bool W, Reg n, RegList reg_list);
```
Comment (lioncash) [inline src/frontend/A32/translate/impl/translate_thumb.h]: We don't use tabs for indentation.

File: tests/A32/fuzz_thumb.cpp [hunk 27]
```diff
@@ -374,6 +374,35 @@ TEST_CASE("Fuzz Thumb2 instructions set 1", "[JitX64][Thumb2]") {
         Thumb32InstGen("11110i01110Snnnn0kkkddddmmmmmmmm", // RSB (imm)
                  [](u32 inst){ return Common::Bits<8, 11>(inst) != 0b1111 && Common::Bits<16, 19>(inst) != 0b1111; }),
             // R15 is UNPREDICTABLE
+		Thumb32InstGen("1110100xx0W0nnnn0r0rrrrrrrrrrrrr", // STMIA / STMDB
```
Comment (lioncash) [inline tests/A32/fuzz_thumb.cpp]: Again, we don't use tabs.

File: src/frontend/A32/translate/impl/helper.cpp (hunk 23 above)
Comment (sunho) [reply src/frontend/A32/translate/impl/helper.cpp]: Should be fixed by https://github.com/MerryMage/dynarmic/pull/552/commits/ec51dc194a3bce8d788ecb4cffbb08892ad54604

File: src/frontend/A32/translate/helper.h (hunk 22 above)
Comment (sunho) [reply src/frontend/A32/translate/helper.h]: Done!

File: tests/A32/fuzz_thumb.cpp [hunk 28]
```diff
@@ -39,13 +40,17 @@ static A32::UserConfig GetUserConfig(ThumbTestEnv* testenv) {
...
+
+template <typename InstructionType>
 struct ThumbInstGen final {
 public:
-    ThumbInstGen(const char* format, std::function<bool(u16)> is_valid = [](u16){ return true; }) : is_valid(is_valid) {
-        REQUIRE(strlen(format) == 16);
-
-        for (int i = 0; i < 16; i++) {
-            const u16 bit = 1 << (15 - i);
+
```
Comment (VelocityRa) [inline tests/A32/fuzz_thumb.cpp:48]: nit: remove newline

**Comment by zhkl0228** [comment]
`TEST_CASE("thumb2: POP", "[thumb2]") {
    ThumbTestEnv test_env;
    Dynarmic::A32::Jit jit{GetUserConfig(&test_env)};
    test_env.code_mem = {
            0xf1b0, 0x0f00, // cmp.w r0, #0
            0xbf04, // itt eq
            0x4629, // mov r1, r5
            0xe8bd, 0x8001, // pop.w {r0, pc}
            0xe7fe, 0xe7fe, // b #0
            0x000d, 0x0000
    };

    jit.Regs()[0] = 0;
    jit.Regs()[5] = 5;
    jit.Regs()[13] = 12; // SP
    jit.Regs()[15] = 0; // PC = 0
    jit.SetCpsr(0x00000030); // Thumb, User-mode

    test_env.ticks_left = 4;
    jit.Run();

    REQUIRE(jit.Regs()[0] == 0xe7fee7fe);
    REQUIRE(jit.Regs()[1] == 5);
    REQUIRE(jit.Regs()[13] == 20);
    REQUIRE(jit.Regs()[15] == 12);
    REQUIRE(jit.Cpsr() == 0x60000030);
}

Test result: dynarmic assertion failed: false
Message: ExceptionRaised(00000008)libc++abi.dylib: terminating`

**Comment by MerryMage** [comment]
Thanks for this PR and kicking my butt into gear.

On `master`, we've reimplemented the IT instruction in a different manner to this PR and also provided a different test framework more inline with the A64 frontent.

I will try to cherry-pick commits from this PR where possible, but I'm closing this PR due to the above.

**Comment by sunho** [comment]
I'm excited to see thumb2 support being added to dynarmic. We'll try dynarmic in vita3k as soon as thumb2 works are finished here. Good luck!

**Pull request closed without merging**